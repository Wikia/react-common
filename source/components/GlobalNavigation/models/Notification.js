import merge from 'lodash/merge';
import get from 'lodash/get';

import { convertToTimestamp } from '../utils/isoTime';

import { notificationTypes } from './notificationTypes';

const defaultProps = {
    title: null,
    snippet: null,
    timestamp: null,
    communityName: null,
    type: null,
    isUnread: false,
    totalUniqueActors: 1,
    latestActors: null,
    uri: null,
};

function mapData(notificationData) {
    return {
        title: get(notificationData, 'refersTo.title'),
        snippet: get(notificationData, 'refersTo.snippet'),
        uri: get(notificationData, 'refersTo.uri'),
        latestEventUri: get(notificationData, 'events.latestEvent.uri'),
        timestamp: convertToTimestamp(get(notificationData, 'events.latestEvent.when')),
        communityName: get(notificationData, 'community.name'),
        communityId: get(notificationData, 'community.id'),
        isUnread: notificationData.read === false,
        totalUniqueActors: get(notificationData, 'events.totalUniqueActors'),
        latestActors: createActors(get(notificationData, 'events.latestActors')),
        type: getNotificationType(notificationData),
    };
}

function createActors(actors) {
    return actors.map(actor => merge({}, actor, { profileUrl: getProfileUrl(actor.username) }));
}

function getProfileUrl(name) {
    return `${window.location.origin}/wiki/User:${name}`;
}

function getNotificationType(apiData) {
    if (apiData.type === 'upvote-notification') {
        if (apiData.refersTo.type === 'discussion-post') {
            return notificationTypes.discussionUpvoteReply;
        }

        return notificationTypes.discussionUpvotePost;
    }

    if (apiData.type === 'replies-notification') {
        return notificationTypes.discussionReply;
    }

    if (apiData.type === 'announcement-notification') {
        return notificationTypes.announcement;
    }

    return null;
}

class Notification {
    constructor(data) {
        Object.keys(data).forEach((key) => {
            this[key] = data[key];
        });
    }

    static build(notificationData) {
        return new this(merge({}, defaultProps, mapData(notificationData)));
    }
}

export default Notification;
