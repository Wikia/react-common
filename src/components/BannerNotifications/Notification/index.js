import React from 'react';
import PropTypes from 'prop-types';

function getIcon(type) {
  switch (type) {
    case ('alert'):
      return '#wds-icons-error-small';
    case ('warning'):
      return '#wds-icons-alert-small';
    case ('success'):
      return '#wds-icons-checkmark-circle-small';
    default:
      return '#wds-icons-flag-small';
  }
}

function getClassName(type) {
  switch (type) {
    case ('alert'):
      return 'wds-alert';
    case ('warning'):
      return 'wds-warning';
    case ('success'):
      return 'wds-success';
    default:
      return 'wds-message';
  }
}

const Notification = ({type, message, onClose}) => (
  <div className={`wds-banner-notification ${getClassName(type)}`}>
    <div className="wds-banner-notification__icon">
      <svg className="wds-icon wds-icon-small">
        <use xmlnsXlink="http://www.w3.org/1999/xlink" xlinkHref={getIcon(type)} />
      </svg>
    </div>
    <span className="wds-banner-notification__text">{message}</span>
    <svg className="wds-icon wds-icon-tiny wds-banner-notification__close" onClick={onClose}>
      <use xmlnsXlink="http://www.w3.org/1999/xlink" xlinkHref="#wds-icons-cross-tiny" />
    </svg>
  </div>
);

Notification.propTypes = {
  type: PropTypes.oneOf(['alert', 'warning', 'success', 'message']).isRequired,
  message: PropTypes.string.isRequired,
  onClose: PropTypes.func.isRequired,
};

export default Notification;
