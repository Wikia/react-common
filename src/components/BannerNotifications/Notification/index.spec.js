import React from 'react';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import sinon from 'sinon';

import Notification from './index';

const noop = () => {};

/* eslint-disable no-alert */
test('Notification renders correctly', () => {
  let component = renderer.create(
    <Notification type={'message'} message={'lorem ipsum - messge'} onClose={noop} />,
  );
  expect(component.toJSON()).toMatchSnapshot();

  component = renderer.create(
    <Notification type={'success'} message={'lorem ipsum - success'} onClose={noop} />,
  );
  expect(component.toJSON()).toMatchSnapshot();

  component = renderer.create(
    <Notification type={'warning'} message={'lorem ipsum - warning'} onClose={noop} />,
  );
  expect(component.toJSON()).toMatchSnapshot();

  component = renderer.create(
    <Notification type={'alert'} message={'lorem ipsum - alert'} onClose={noop} />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Notification onClose hander is invoked', () => {
  const mockOnClick = sinon.spy();
  const wrapper = shallow(
    <Notification type={'message'} message={'lorem ipsum - messge'} onClose={mockOnClick} />
  );

  wrapper.find('.wds-banner-notification__close').simulate('click');

  expect(mockOnClick.calledOnce).toBe(true);
});
