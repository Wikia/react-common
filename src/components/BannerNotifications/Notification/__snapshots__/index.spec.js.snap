// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification renders correctly 1`] = `
<div
  className="wds-banner-notification wds-message"
>
  <div
    className="wds-banner-notification__icon"
  >
    <svg
      className="wds-icon wds-icon-small"
    >
      <use
        xlinkHref="#wds-icons-flag-small"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </div>
  <span
    className="wds-banner-notification__text"
  >
    lorem ipsum - messge
  </span>
  <svg
    className="wds-icon wds-icon-tiny wds-banner-notification__close"
    onClick={[Function]}
  >
    <use
      xlinkHref="#wds-icons-cross-tiny"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</div>
`;

exports[`Notification renders correctly 2`] = `
<div
  className="wds-banner-notification wds-success"
>
  <div
    className="wds-banner-notification__icon"
  >
    <svg
      className="wds-icon wds-icon-small"
    >
      <use
        xlinkHref="#wds-icons-checkmark-circle-small"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </div>
  <span
    className="wds-banner-notification__text"
  >
    lorem ipsum - success
  </span>
  <svg
    className="wds-icon wds-icon-tiny wds-banner-notification__close"
    onClick={[Function]}
  >
    <use
      xlinkHref="#wds-icons-cross-tiny"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</div>
`;

exports[`Notification renders correctly 3`] = `
<div
  className="wds-banner-notification wds-warning"
>
  <div
    className="wds-banner-notification__icon"
  >
    <svg
      className="wds-icon wds-icon-small"
    >
      <use
        xlinkHref="#wds-icons-alert-small"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </div>
  <span
    className="wds-banner-notification__text"
  >
    lorem ipsum - warning
  </span>
  <svg
    className="wds-icon wds-icon-tiny wds-banner-notification__close"
    onClick={[Function]}
  >
    <use
      xlinkHref="#wds-icons-cross-tiny"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</div>
`;

exports[`Notification renders correctly 4`] = `
<div
  className="wds-banner-notification wds-alert"
>
  <div
    className="wds-banner-notification__icon"
  >
    <svg
      className="wds-icon wds-icon-small"
    >
      <use
        xlinkHref="#wds-icons-error-small"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
  </div>
  <span
    className="wds-banner-notification__text"
  >
    lorem ipsum - alert
  </span>
  <svg
    className="wds-icon wds-icon-tiny wds-banner-notification__close"
    onClick={[Function]}
  >
    <use
      xlinkHref="#wds-icons-cross-tiny"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</div>
`;
